### PR TITLE
Ctc unit test

### PR DESCRIPTION
**Pull request by anjaligopi**
Add unit test for CTC model
Add calculateWER() to compute WER in computeLoss
Rebased to pad-batch-dimension

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -4,7 +4,6 @@ channels:
   - conda-forge
   - defaults
 dependencies:
-  - _libgcc_mutex=0.1=conda_forge
```
Comment (galv) [inline environment.yml:7]: To clarify, did these fail to download when you installed the conda environment? I've seen that happen before. In that case, it's probably fine to remove them and come back to this later if necessary.

File: lingvo/tasks/asr/ctc_model.py [hunk 2]
```diff
@@ -138,62 +140,118 @@ def __init__(self, params):
...
-        #   # Ugh, I hate transposes like this!
-        #   rnn_out, rnn_padding = self.stacking.FProp(
-        #       tf.transpose(rnn_out, [1, 0, 2]),
-        #       tf.transpose(rnn_padding, [1, 0, 2]))
-        #   rnn_out = tf.transpose(rnn_out, [1, 0, 2])
-        #   rnn_padding = tf.transpose(rnn_padding, [1, 0, 2])
-
-      rnn_out = self.project_to_vocab_size(rnn_out)
-      rnn_out *= (1.0 - rnn_padding)
+      p = self.params
```
Comment (galv) [inline lingvo/tasks/asr/ctc_model.py:228]: As a word of advise, you should set your editor to use two space indentations for this project. Google uses 2 space indents in its projects for whatever reason, so it's best to stick with this.

File: test [hunk 3]
```diff
@@ -0,0 +1 @@
+blah blah
```
Comment (galv) [inline test:1]: Delete this

File: test_gcloud_storage.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from gcloud import storage
```
Comment (galv) [inline test_gcloud_storage.py:1]: Try not to commit personal working files like these.

**Review by galv — COMMENTED** [review]
Alright looks good. I'll try to actually run the code tomorrow-ish.

File: environment.yml (hunk 1 above)
Comment (anjaligopi) [reply environment.yml:7]: yeah, conda env creation didn't work until I removed these things. I was still able to build lingvo successfully w/o explicitly having these (these may have been installed as dependencies of other things).

File: lingvo/tasks/asr/ctc_model.py (hunk 2 above)
Comment (anjaligopi) [reply lingvo/tasks/asr/ctc_model.py:228]: oh thanks for this tip, didn't realize

File: test (hunk 3 above)
Comment (anjaligopi) [reply test:1]: these are my bad, I didnt clean up a lot of other things to in the commit, sorry about that :( Really appreciate the detailed review, thanks!

File: lingvo/tasks/asr/ctc_model.py [hunk 5]
```diff
@@ -138,62 +140,118 @@ def __init__(self, params):
...
+              tf.squeeze(output_batch.encoder_outputs_padding, 2), 0
+          ),
+      )
+
+      dec = tf.sparse_to_dense(
+          decoded.indices, decoded.dense_shape, decoded.values, default_value=73
+      )
+
+      hyp_str = self.input_generator.IdsToStrings(
+          tf.cast(dec, tf.int32), tf.shape(dec)
```
Comment (galv) [inline lingvo/tasks/asr/ctc_model.py]: tf.shape() isn't right here. The "shapes" as the second argument to IdsToStrings is the length of each sequence in the batch, not the shape of the tensor (which is essentially length along each dimension). We need to convert the indices array into lengths. Probably a reduce-by-key (where the reduce is the max operation) would do it. I'm trying to figure this out.

File: lingvo/tasks/asr/ctc_model.py [hunk 6]
```diff
@@ -138,62 +140,118 @@ def __init__(self, params):
...
+  def _CalculateWER(self, input_batch, output_batch):
+      (decoded,), neg_sum_logits = tf.nn.ctc_greedy_decoder(
+          output_batch.encoder_outputs,
+          py_utils.LengthsFromBitMask(
+              tf.squeeze(output_batch.encoder_outputs_padding, 2), 0
+          ),
+      )
+
+      dec = tf.sparse_to_dense(
+          decoded.indices, decoded.dense_shape, decoded.values, default_value=73
```
Comment (galv) [inline lingvo/tasks/asr/ctc_model.py]: No need to  use 73 as the default value here. We just need to get the padding right, and the default values can be arbitrary.

File: lingvo/tasks/asr/ctc_model.py (hunk 5 above)
Comment (galv) [reply lingvo/tasks/asr/ctc_model.py]: Either reduce-by-key or tf.scatter_nd.
